src/viewport.js

```diff
@@ -1522,15 +1522,53 @@ $.Viewport.prototype = {
     },
 
     /**
-     * Toggle flip state and demands a new drawing on navigator and viewer objects.
-    */
+     * Toggles flip state and demands a new drawing on navigator and viewer objects.
+     * @function
+     * @return {OpenSeadragon.Viewport} Chainable.
+     */
     toggleFlip: function() {
-      this.flipped = !this.flipped;
+      this.setFlip(!this.getFlip());
       if(this.viewer.navigator){
-        this.viewer.navigator.toggleFlip();
+        this.viewer.navigator.setFlip();
       }
-      this.viewer._forceRedraw = !this.viewer._forceRedraw;
       this.viewer.forceRedraw();
+      return this;
+    },
+
+
+    /**
+     * Gets flip state stored on viewport.
+     * @function
+     * @return {Boolean} Flip state.
+     */
+    getFlip: function() {
+      return this.flipped;
+    },
+
+
+    /**
+     * Sets flip state according to the state input argument.
+     * @function
+     * @param {Boolean} state - Flip state to set.
+     * @return {OpenSeadragon.Viewport} Chainable.
+     */
+    setFlip: function( state ) {
+      if ( this.flipped != state ) {
```

Comment:
The event should be raised at the very end of the function only after everything else has been done. Otherwise the event might trigger something that would query the flipped state and find it wrong.

Reply:
I did that to avoid a ugly if/else in this case but you are 100% correct. Done.